File: src/lexer.rs
use token::{Token , TokenType};
use crate::errors::lexer_errors::LexerError;
use crate::errors::lexer_errors::LexerError::{InvalidNumber, UnexpectedCharacter};

pub mod token;

pub struct Lexer {
    index: usize,
    line: usize,
    start: usize,
    text: Vec<char>
}

impl Lexer {
    pub fn new () -> Self {
        Lexer{
            index: 0,
            line: 1,
            start: 0,
            text: vec![]
        }
    }

    pub fn parse(&mut self,text:&String) ->Result<Vec<Token>, LexerError >{
        self.text = text.chars().collect();
        self.line = 1;
        self.index = 0;
        let mut res = vec![];

        while let Some(&chr) = self.text.get(self.index){
            match chr {
                ' ' | '\t' | '\r' => {
                    self.index += 1;
                    self.start += 1;
                },
                '\n' => {
                    self.line += 1;
                    self.index += 1;
                    self.start = 0;
                }
                ident if Self::is_allowed_ident(ident) => {
                    let name = self.parse_ident();
                    res.push(name)
                }
                digit if digit.is_ascii_digit() => {
                    let num = self.parse_number()?;
                    res.push(num);
                }
                '+' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::PlusEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Plus, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '-' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::MinusEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Minus, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '*' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::StarEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Star, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '/' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::SlashEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Slash, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '%' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::PercentEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Percent, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '!' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::NotEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Not, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '=' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::DoubleEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Equal, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '<' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::LessEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else if self.text.get(self.index + 1) == Some(&'<') {
                        if self.text.get(self.index + 2) == Some(&'=') {
                            res.push(Token::new(TokenType::DoubleLessEqual, self.index, self.start, 3, self.line));
                            self.index += 3;
                            self.start += 3;
                        } else {
                            res.push(Token::new(TokenType::DoubleLess, self.index, self.start, 2, self.line));
                            self.index += 2;
                            self.start += 2;
                        }
                    } else {
                        res.push(Token::new(TokenType::Less, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '>' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::GreaterEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else if self.text.get(self.index + 1) == Some(&'>') {
                        if self.text.get(self.index + 2) == Some(&'=') {
                            res.push(Token::new(TokenType::DoubleGreaterEqual, self.index, self.start, 3, self.line));
                            self.index += 3;
                            self.start += 3;
                        } else {
                            res.push(Token::new(TokenType::DoubleGreater, self.index, self.start, 2, self.line));
                            self.index += 2;
                            self.start += 2;
                        }
                    } else {
                        res.push(Token::new(TokenType::Greater, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '~' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::TildeEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Tilde, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '&' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::AmpersandEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Ampersand, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '|' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::PipeEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Pipe, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '^' => {
                    if self.text.get(self.index + 1) == Some(&'=') {
                        res.push(Token::new(TokenType::CaretEqual, self.index, self.start, 2, self.line));
                        self.index += 2;
                        self.start += 2;
                    } else {
                        res.push(Token::new(TokenType::Caret, self.index, self.start, 1, self.line));
                        self.index += 1;
                        self.start += 1;
                    }
                }
                '.' => {
                    res.push(Token::new(TokenType::Dot, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                ',' => {
                    res.push(Token::new(TokenType::Comma, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                '(' => {
                    res.push(Token::new(TokenType::LeftRBracket, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                ')' => {
                    res.push(Token::new(TokenType::RightRBracket, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                '{' => {
                    res.push(Token::new(TokenType::LeftCBracket, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                '}' => {
                    res.push(Token::new(TokenType::RightCBracket, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                '[' => {
                    res.push(Token::new(TokenType::LeftSBracket, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                ']' => {
                    res.push(Token::new(TokenType::RightSBracket, self.index, self.start, 1, self.line));
                    self.index += 1;
                    self.start += 1;
                }
                _ => return Err(UnexpectedCharacter {line: self.line, index: self.index})
            }
        }
        Ok(res)
    }
    
    fn parse_number(&mut self) -> Result<Token, LexerError> {
        let start = self.index;
        while let Some(&chr) = self.text.get(self.index) {
            if chr.is_ascii_digit(){
                self.index += 1;
            }
            else{
                break;
            }
        }

        if let Some(&chr) = self.text.get(self.index) {
            if Self::is_allowed_ident(chr){
                return Err(InvalidNumber{line: self.line,index: start, len: self.index - start});
            }
        }

        Ok(Token::new(TokenType::Integer, start, self.start, self.index - start, self.line))
    }

    fn parse_ident(&mut self) -> Token {
        let start = self.index;
        while let Some(&chr) = self.text.get(self.index) {
            if Self::is_allowed_ident(chr) || chr.is_ascii_digit(){
                self.index += 1;
            }
            else{
                break;
            }
        }

        let word:String = self.text[start..self.index].iter().collect();
        let token_type = match &*word {
            "if" => TokenType::If,
            "else" => TokenType::Else,
            "elif" => TokenType::Elif,
            "while" => TokenType::While,
            "fn" => TokenType::Function,
            "struct" => TokenType::Struct,
            "union" => TokenType::Union,
            "var" => TokenType::Var,
            "const" => TokenType::Const,
            "and" => TokenType::And,
            "or" => TokenType::Or,
            "namespace" => TokenType::Namespace,
            "use" => TokenType::Use,
            "return" => TokenType::Return,
            "break" => TokenType::Break,
            "continue" => TokenType::Continue,
            "true" => TokenType::True,
            "false" => TokenType::False,
            _=> TokenType::Ident
        };

        Token::new(token_type , start, self.start, self.index - start, self.line)
    }

    fn is_allowed_ident(chr:char) -> bool{
        chr.is_ascii_alphabetic() ||
            chr == '_' ||
            chr as u32 > 127
    }
}

File: src/compiler/lexer.cpp
#include "lexer.h"

#include <iostream>
#include <locale>

std::string tokenTypeToString(TokenType type) {
    switch (type) {
        case EOF_: return "EOF_";
        case WORD: return "WORD";
        case NUMBER: return "NUMBER";
        case TRUE: return "TRUE";
        case FALSE: return "FALSE";
        case PLUS: return "PLUS";
        case MINUS: return "MINUS";
        case STAR: return "STAR";
        case SLASH: return "SLASH";
        case LEFT_PARENTHESIS: return "LEFT_PARENTHESIS";
        case RIGHT_PARENTHESIS: return "RIGHT_PARENTHESIS";
        case LEFT_BRACE: return "LEFT_BRACE";
        case RIGHT_BRACE: return "RIGHT_BRACE";
        case EQUAL: return "EQUAL";
        case DOUBLE_EQUAL: return "DOUBLE_EQUAL";
        case IF: return "IF";
        case ELSE: return "ELSE";
        case AND: return "AND";
        case OR: return "OR";
        case LESS: return "LESS";
        case LESS_EQUAL: return "LESS_EQUAL";
        case GREATER: return "GREATER";
        case GREATER_EQUAL: return "GREATER_EQUAL";
        case NOT: return "NOT";
        case NOT_EQUAL: return "NOT_EQUAL";
        case WHILE: return "WHILE";
        case BREAK: return "BREAK";
        case CONTINUE: return "CONTINUE";
        case PLUS_EQUAL: return "PLUS_EQUAL";
        case MINUS_EQUAL: return "MINUS_EQUAL";
        case STAR_EQUAL: return "STAR_EQUAL";
        case SLASH_EQUAL: return "SLASH_EQUAL";
        default: return "UNKNOWN";
    }
}

std::ostream & operator<<(std::ostream &os, const Token &token) {
    os <<"{ "<<"type: "<<tokenTypeToString(token.type)<<" start: "<<token.start<<" end: "
    <<token.end<<" line: "<<token.line<<" }";
    return os;
}

bool Lexer::is_end() {
    return index >= chars.size();
}

Token Lexer::make_token(TokenType type) {
    return Token{type,start, index, line};
}

bool Lexer::match(char c) {
    if (is_end())
        return false;
    if (chars[index] == c) {
        index++;
        return true;
    }

    return false;
}

void Lexer::skip_ws() {
    while (!is_end()) {
        switch (chars[index]) {
            case ' ':
            case '\t':
            case '\r':
                index++;
                break;
            case '\n':
                index++;
                line++;
                break;
            default: return;;
        }
    }
}

Token Lexer::number() {
    while (!is_end() && isdigit(chars[index])) {
        index++;
    }

    if (!is_end() && chars[index] == '.' && isdigit(chars[index + 1])) {
        index++;
        while (!is_end() && isdigit(chars[index])) {
            index++;
        }
    }

    return make_token(TokenType::NUMBER);
}

Token Lexer::ident() {
    while (!is_end() &&
        (isalpha(chars[index]) || isdigit(chars[index]))
        ) {
        index++;
    }

    TokenType type;
    auto word = chars.substr(start, index - start);

        if(word ==  "ճիշտ")
            type = TokenType::TRUE;
        else if(word == "եթե")
            type = TokenType::IF;
        else if(word == "ուրիշ")
            type = TokenType::ELSE;
        else if(word == "կեխծ")
            type = TokenType::FALSE;
        else if(word == "և")
            type = TokenType::AND;
        else if(word == "կամ")
            type = TokenType::OR;
        else if(word == "ցիկլ")
            type = TokenType::WHILE;
        else if(word == "կայնի")
            type = TokenType::BREAK;
        else if(word == "շարունակի")
            type = TokenType::CONTINUE;
        else
            type = TokenType::WORD;

        return make_token(type);

}

Lexer::Lexer(std::string& chars):chars(chars), start(0), index(0), line(1) {}

std::vector<Token> Lexer::parse() {
    std::vector<Token> tokens;
    while (true) {
        skip_ws();

        if (is_end())
            break;
        start = index;

        switch (chars[index++]) {
            case '+':
                if (match('='))
                    tokens.push_back(make_token(TokenType::PLUS_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::PLUS));
                break;

            case '-':
                if (match('='))
                    tokens.push_back(make_token(TokenType::MINUS_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::MINUS));
                break;
            case '*':
                if (match('='))
                    tokens.push_back(make_token(TokenType::STAR_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::STAR));
                break;
            case '/':
                if (match('='))
                    tokens.push_back(make_token(TokenType::SLASH_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::SLASH));
                break;
            case '(':
                tokens.push_back(make_token(TokenType::LEFT_PARENTHESIS));
                break;
            case ')':
                tokens.push_back(make_token(TokenType::RIGHT_PARENTHESIS));
                break;
            case '=':
                if (match('='))
                    tokens.push_back(make_token(TokenType::DOUBLE_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::EQUAL));
                break;
            case '!':
                if (match('='))
                    tokens.push_back(make_token(TokenType::NOT_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::NOT));
                break;
            case '<':
                if (match('='))
                    tokens.push_back(make_token(TokenType::LESS_EQUAL));
                else
                    tokens.push_back(make_token(TokenType::LESS));
                break;
            case '{':
                tokens.push_back(make_token(TokenType::LEFT_BRACE));
                break;
            case '}':
                tokens.push_back(make_token(TokenType::RIGHT_BRACE));
                break;
            default:
                char chr = chars[index-1];
                if (std::isdigit(chr) ) {
                    auto token = number();
                    tokens.push_back(token);
                }
                else if (std::isalpha(chr) || chr == '_') {
                    auto token = ident();
                    tokens.push_back(token);
                }
                else {
                    //throw
                    std::cerr<<"EROR";
                }
        }
    }
    tokens.push_back(make_token(TokenType::EOF_));
    return tokens;
}
